protobuf/lib/src/protobuf/json.dart

```diff
@@ -24,22 +24,22 @@ Map<String, dynamic> _writeToJsonMap(_FieldSet fs) {
         return fieldValue;
       case PbFieldType._FLOAT_BIT:
       case PbFieldType._DOUBLE_BIT:
-        final value = fieldValue as double;
+        final value = downcastUnchecked<double>(fieldValue);
         if (value.isNaN) {
           return _nan;
         }
         if (value.isInfinite) {
           return value.isNegative ? _negativeInfinity : _infinity;
         }
-        if (fieldValue.toInt() == fieldValue) {
-          return fieldValue.toInt();
+        if (value.toInt() == fieldValue) {
```

Comment:
drive by comment: will this behave differently in wasm than in JS? Do you need `value.toInt().toDouble() == value` or something else to ensure we are in the same vaue domain?